### PR TITLE
Add in bors

### PR DESCRIPTION
**Pull request by cb22**
Add in [bors](https://bors.tech/) (it's already installed on the repo) for merge trains / queuing. We also bump the Node client's `package.json` version to fix that failing check here.

### CODE REVIEW

**Comment by cb22** [comment]
bors try

**Comment by matklad** [comment]
:man_dancing: :tada: :woman_dancing: :tada: :man_dancing:

**Comment by cb22** [comment]
@matklad was just about to ask if you have any thoughts! :smile:

**Comment by cb22** [comment]
bors try

File: .github/workflows/clients-dotnet-build.yml
Comment (matklad) [inline .github/workflows/clients-dotnet-build.yml:148]: These are basically grouping steps to avoid listing a million of entries in bors.toml, right? It feels like a good hack here, though naively it seems that having so much CI jobs is the problem, I wonder if we can compress that somehow, using matricies or some such. Definitely out of scope for the PR.

File: .github/workflows/clients-java-build.yml
Comment (matklad) [inline .github/workflows/clients-java-build.yml:55]: Out of scope for this PR, but maybe we should have two different yamls:

* `ci.yml` which runs tests, gates bors, and is run on every branch (technically, with bors it isn't necessary to run tests on master, as that's redundant, but that's also a good way to catch flakes)
* `release.yml` which does the release, and is run on master branch only. 

Within bors paradigm, clearly separating "check that gate master branch" from "everything else we do automatically with known-good master branch" makes sense philosophically. 

Don't think there are any direct practical benefits though.

File: bors.toml
Comment (matklad) [inline bors.toml:2]: I'd maybe

```suggestion
status = [
```

and rely on bors mechanism for approvals. 

I think that bors's way (r+ comments) is better. Sepcifically, reviewer, after they are satisfied with the code, leaves the `bors r+` comment, and that is considered an approval that triggers the merge, regardless of the presence of GH-level approval. This is better because: 

* It removes the ambiguity for "who presses the merge button". Reviewer does that, by explicitly leaving the `r+` comment. So, ultimately it's the reviewer who is on the hook to make  sure that the code we merge is good, which I think is a good psychological split of responsibilities. This also dovetails with idea I've heard from @jorangreef  that we deliberately have only a single reviewer to avoid bystander problem
* It removes ambiguity of what exactly is approved: `r+` always applies to the current commit hash.
* It makes "LGTM, but please this one small nit before merging" explicit: there's `bors d+` command which means exactly "the PR author can now approve their own change".

File: .github/workflows/clients-dotnet-build.yml
Comment (cb22) [reply .github/workflows/clients-dotnet-build.yml:148]: Yes, exactly for that reason. Their getting started guide recommends a collecting step if you're using matrix builds too.

+1 on cleaning up the steps in a separate PR - the client ones aren't too bad, but `test_and_build.yml` has a lot of steps that could really just be a test suite (eg, all the `fuzz_`)

Comment (matklad) [reply .github/workflows/clients-dotnet-build.yml:148]: Actually, it seems there's perhaps a better way to do that? You can use regexes in bors: 


>Patterns are written using SQL LIKE syntax. For example, this status directive will match any CI job that starts with ci/gitlab/:

```
status = [ 'ci/gitlab/%' ]
```

File: .github/workflows/clients-java-build.yml
Comment (cb22) [reply .github/workflows/clients-java-build.yml:55]: I agree - bors aside, we currently gate releasing in different ways depending on the case (eg, in the node / go clients a whole job is gated behind `github.ref == 'refs/heads/main'` whereas the other clients do a different step within a job depending.)

The `build-and-push` step does it as a flag, which gave me a mini-fright the first time I skimmed over the config and didn't see it being conditional on `main` :smile:

File: bors.toml
Comment (matklad) [reply bors.toml:2]: But also, I'd be totally up for merging this as is, and tweaking config later!

Comment (cb22) [reply bors.toml:2]: I like how that sounds - I went with `required_approvals` just to keep change down to a minimum. Deferring to @sentientwaffle here.

File: .github/workflows/clients-dotnet-build.yml
Comment (matklad) [reply .github/workflows/clients-dotnet-build.yml:148]: Ah, if the docs recommend collecting step, that's the way to go!

File: bors.toml
Comment (sentientwaffle) [reply bors.toml:2]: Agreed, that sounds more convenient — being able to say "PR author can approve their own change" is perfect.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**
